### ISSUE TITLE
[FR] Architecture: method could be static by @rentalhost

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#204]
Origin: https://bitbucket.org/kalessil/phpinspectionsea/issues/113
Author: @rentalhost 

Can be simple check if a method can be static. Basically, just check if in some point it refer to `$this`.

Example (_could be static_):

```php
function test() {
    return 'hello';
}
```

Example (_can't be static_, `$this` is used to refer same object):

```php
function test() {
    return $this->name;
}
```

**Comment by rentalhost** [comment]
@funivan What do you mean by _confusing_ emoji? Do you think that could be some expection on this case?

**Comment by funivan** [comment]
Static methods is hard to test. So in most cases it is better to have some performance downgrade and use non static methods. ;)

**Comment by kalessil** [comment]
Hm, I thought we could spot such methods and suggest decoupling them from the class.
If we have a method which can be static, it most probably not needed in the class at all.

Or?

**Comment by funivan** [comment]
>  If we have a method which can be static, it most probably not needed in the class at all.

Maybe i need to see more examples ;)

**Comment by kalessil** [comment]
I need a prototype to spot examples =)

**Comment by Fleshgrinder** [comment]
> Static methods is hard to test. 
> — @funivan 

This is generally true, but not an argument against static methods in general; it highly depends. In case it’s a _pure_ function without state and side effects, and it is used to protect the invariant of a class, nothing is wrong. More important, it is not a dependency that should be broken! The example in this issue is once more very simple, but it is a _pure_ function, no side effects, no hard wiring, no strategies, there is nothing wrong about it. The question would be, what is the string that is being returned by this _pure_ function going to be used for. Would I actually need to exchange it, to avoid other side effects? This is pretty much impossible to answer without more context.

**TL;DR** there is no absolute answer to the problem at hand without context.
